### ISSUE TITLE
Section for non-visible controls

### ISSUE THREAD

**Issue by bclothier** [WaynePhillipsEA/twinbasic#911]
**Is your feature request related to a problem? Please describe.**
Now that we have a form and ActiveX support, I would like to make a proposal in regards to non-visible controls. It never felt right to be putting those on the form itself, though I understand the rationale for doing that. Even in Access where we deal with visible controls that may be invisible at run-time, this presents issues which boils down to making the following choices:

1) do I make the form layout tight and overlay the non-visible controls on other controls
vs.
2) do I just move all non-visible controls into blank area of the form (and if needed, resize a bit to make them visible).

**Describe the solution you'd like**

Neither are great choices. So I'd like a 3rd choice: having a dedicated section where all non-visible controls can be put, and arranged in a grid and is all easily visible. That section would have "infinite" area so there is no need to overlay anything so it's easy to just scroll and see what non-visible controls there are on the current form. That section would also eliminate the need to maintain the `Top`,`Left`, `Height` and `Width` properties from the property sheet since they now would be irrelevant. 

That would allow us to make the form layout tight as we want without worrying about cluttering it so much that it's hard to see all the controls, both visible and non-visible.

**Describe alternatives you've considered**
Status quo.

**Comment by WaynePhillipsEA** [comment]
I like this idea.   I wonder how we might implement it in the UI.  Would we display a bar (e.g. at the top of the designer) containing the non-visible controls once we first add a non-visual control to a form, or would we want it as a permanent section of the designer?

**Comment by jpbro** [comment]
I like this idea too -I've never been a fan of non-visual controls on forms (even as far back as my early days learning VB - I recall that they were confusing from the start. Not sure what the best solution is, but in VB6 we have "View Code" and "View Object" buttons, so perhaps we could also have a "View <WhateverThey'reCalled>" button?

**Comment by camomille22003** [comment]
How to drop an invisible button on a Frame for anchor it if it is not present on the Form ?

**Comment by WaynePhillipsEA** [comment]
@camomille22003 we are talking about non-visual controls, for example a timer.   I will adjust the post to say non-visual rather than non-visible for clarity.

**Comment by bclothier** [comment]
Apologies for poor wording. 

I envisioned a section among the bottom of the designer but it could be a dockable panel. For example, SSIS designer has their connection managers in a section on the bottom:
![image](https://user-images.githubusercontent.com/2367644/174283157-82a4ff18-e244-4a01-9253-56da7d686494.png)

Because we need to deal with back-compat, I envision having a toggle button. If untoggled, the non-visual controls are placed normally on the control, but if toggled, they all get moved to the section and are automatically laid out to a grid.

**Comment by bclothier** [comment]
One more consideration -- would we also want to allow moving invisible controls to the section for non-visual controls? I don't know how common it may be in VB6 to make use of invisible controls as placeholder for data (it's pretty common in Access --- likely as a consequence of data binding their forms/controls does).

**Comment by Kr00l** [comment]
All non-visual control do draw an image of itself on design time and resize differently (e.g. timer vs ImageList)
So how to accomodate that in an off-form grid section?

**Comment by bclothier** [comment]
Similar to how SSIS connection managers -- the size of the connection manager is variable because it contains the name of the connection manager, which can be short or long or in between. If it overlays, SSIS automatically snaps it to a place so that none overlays. Again, I might be using poor choice of words as you see in the screenshot above, there is no visible grid but everything is snapped to a variable-size slot on the grid for layout.

**Comment by Kr00l** [comment]
How would the non visual controls be ordered?

**Comment by Kr00l** [comment]
How about a form ide setting which woud allow either form placement or off-form placement.
Because I can imagine that some folks like to have their non-visual controls on the form and "nearby" a control it get's attached to..

About the arranging in the off-form grid.
We could recycle the Left/Top properties (they still exists) and calculate the ordering based on that.

**Comment by bclothier** [comment]
I thought about that but if we re-use those properties, then it'll mess with the layout if they toggle between the off-form section and on-form placement. I would hope that one wouldn't be in habit of toggling between the two view styles but if we allow toggling then we need to ensure that one view doesn't change the other view.

**Comment by Kr00l** [comment]
Yes after a thought Left/Top should be left out. But of course persisted so that the two "views" can be toggled.
So another "property" must be created.

BUT, could that new property be only "virtual". ?Means it's not a real property which can be found on the Extender object but it's rather a sorted list on the Form's JSON serialization.

**Comment by bclothier** [comment]
Yep, that's what I was thinking - it would be only used for the designer's rendering. That way, when the section is enabled, and when someone adds a new visual control, the Top/Left would default to `0` and it'd get added to the end of the visual list on the section with the virtual properties set to represent its position on the new section. If the section supports drag'n'drop and reflowing the controls' position in reaction, then it may not be necessary to show those properties on the property sheet at all since it'd be managed by the section. That would help the user minimize the time spent on fiddling on designer-only property and be more productive.

**Comment by mburns08109** [comment]
+1 on this suggestion from me - if that matters.
